proposals/0279-multiple-trailing-closures.md

```diff
@@ -1,422 +1,369 @@
+
+The labeled trailing closures are associated with the base expression in the same way as the unlabeled trailing closure is today:
 
-## Detailed design
+* if the base expression is a call, they are added as extra arguments to that call;
+* if the base expression is a subscript, they are added as extra index arguments to that subscript,
+* otherwise, an implicit call of the base expression is created using only the trailing closures as arguments.
 
-Changes required to implement new syntax are constrained solely to the parser, namely to the parsing of call arguments.
+The existing trailing-closure feature requires special treatment of the trailing-closure argument by the type checker because of the special power of label omission: the trailing closure can be passed to a parameter that would ordinarily require an argument label.  Currently, the special treatment is specific to the final argument.  Because this proposal still has an unlabeled trailing-closure argument, we have to generalize that treatment to allow label omission at an intermediate argument.
 
-It's possible to do an early syntax transformation which would consider each block to be a regular argument and adjust locations of parentheses if necessary.
-For type checker perspective this means that calls with new syntax would just be regular calls which require no special handling:
+Note that labeled trailing closures are required to match labels with a parameter.  A labeled trailing closure can use the special label `_` to indicate that it matches an unlabeled parameter, but this *only* matches an unlabeled parameter; it does not have the special label-omission powers of the initial unlabeled trailing closure.  For example:
 
 ```swift
-when(2 < 3) {
-  then: { ... }
-  else: { ... }
+func pointFromClosures(x: () -> Int, _ y: () -> Int) -> (Int, Int) { (x,y) }
+pointFromClosures { 10 } _: { 20 }  // Ok
+
+func performAsync(action: @escaping () -> (),
+                  completionOnMainThread: @escaping () -> ()) { ... }
```

Comment:
* Return `-> Void` instead of `-> ()`

```suggestion
func performAsync(
  action: @escaping () -> Void,
  completionOnMainThread: @escaping () -> Void
) {
  ...
}
```